client/post-editor/post-editor.jsx

```diff
@@ -1013,17 +1014,268 @@ export const PostEditor = React.createClass( {
+		 * TinyMCE does some serialization of the content, to fix things like shortcodes, run plugins, prettify
+		 * HTML code and so on. In this process, the bookmark markup gets lost.
+		 *
+		 * If we decide to hook right after the bookmark is added, we can see where the bookmark is in the raw HTML
+		 * in TinyMCE. Unfortunately this state is before the serialization, so any visual markup in the content will
+		 * throw off the positioning.
+		 *
+		 * To avoid this, we insert two custom `span`s that will serve as the markers at the beginning and end of the
+		 * selection.
+		 *
+		 * Why not use TinyMCE's selection API or the DOM API to wrap the contents? Because if we do that, this creates
+		 * a new node, which is inserted in the dom. Now this will be fine, if we worked with fixed selections to
+		 * full nodes. Unfortunately in our case, the user can select whatever they like, which means that the
+		 * selection may start in the middle of one node and end in the middle of a completely different one. If we
+		 * wrap the selection in another node, this will create artifacts in the content.
+		 *
+		 * Using the method below, we insert the custom `span` nodes at the start and at the end of the selection.
+		 * This helps us not break the content and also gives us the option to work with multi-node selections without
+		 * breaking the markup.
+		 */
+		const range = selection.getRangeAt( 0 );
+		const startNode = range.startContainer;
+		const startOffset = range.startOffset;
+		const boundaryRange = range.cloneRange();
+
+		boundaryRange.collapse( false );
+		boundaryRange.insertNode( endElement );
+		boundaryRange.setStart( startNode, startOffset );
+		boundaryRange.collapse( true );
+		boundaryRange.insertNode( startElement );
+
+		range.setStartAfter( startElement );
+		range.setEndBefore( endElement );
+		selection.removeAllRanges();
+		selection.addRange( range );
+
+		/**
+		 * Now the editor's content has the start/end nodes.
+		 *
+		 * Unfortunately the content goes through some more changes after this step, before it gets inserted
+		 * in the `textarea`. This means that we have to do some minor cleanup on our own here.
+		 */
+		const content = editor.getContent()
+			.replace( /(\[\/[^\]]+\]\s*)<p>/g, '$1\n\n' ) // fix short codes
+			.replace( /<p>/g, '' ) // remove `<p>`s, as they don't play any role
+			.replace( /<\/p>/g, '\n\n' ); // replace `</p>` with two new lines.
+
+		const startRegex = new RegExp(
+			'<span[^>]*\\s*class="mce_SELRES_start"[^>]+>\\s*' + selectionID + '[^<]*<\\/span>'
+		);
+
+		const endRegex = new RegExp(
+			'<span[^>]*\\s*class="mce_SELRES_end"[^>]+>\\s*' + selectionID + '[^<]*<\\/span>'
+		);
+
+		// Find the `_start` node
+		const startMatch = content.match( startRegex );
+		if ( ! startMatch ) {
+			return null;
+		}
+
+		const selectionRange = {
+			start: startMatch.index,
+		};
+
+		// Find the `_end` node
+		const endMatch = content.match( endRegex );
```

Comment:
If we bumped `endMatch` up to be above the definition of `selectionRange` are we the able to just do the following?:

```js
const startMatch = ...
if ( startMatch ) ...

const endMatch = ...
if ( endMatch ) ...

const selectionRange = {
  start = startMatch.index,
  end = endMatch.index - startMatch[ 0 ].length,
};
```

Reply:
Tried to do some premature optimizations, but your way is much more readable and sensible. Updated the code!